test/e2e/framework/kubernetes/create-agnhost-statefulset.go

```diff
@@ -0,0 +1,157 @@
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+var ErrLabelMissingFromPod = fmt.Errorf("label missing from pod")
+
+const (
+	AgnhostHTTPPort = 80
+	AgnhostReplicas = 1
+)
+
+type CreateAgnhostStatefulSet struct {
+	AgnhostName        string
+	AgnhostNamespace   string
+	KubeConfigFilePath string
+}
+
+func (c *CreateAgnhostStatefulSet) Run() error {
+	config, err := clientcmd.BuildConfigFromFlags("", c.KubeConfigFilePath)
+	if err != nil {
+		return fmt.Errorf("error building kubeconfig: %w", err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return fmt.Errorf("error creating Kubernetes client: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeoutSeconds*time.Second)
+	defer cancel()
+
+	agnhostStatefulest := c.getAgnhostDeployment()
+
+	err = CreateResource(ctx, agnhostStatefulest, clientset)
+	if err != nil {
+		return fmt.Errorf("error agnhost component: %w", err)
+	}
+
+	selector, exists := agnhostStatefulest.Spec.Selector.MatchLabels["app"]
+	if !exists {
+		return fmt.Errorf("missing label \"app=%s\" from agnhost statefulset: %w", c.AgnhostName, ErrLabelMissingFromPod)
+	}
+
+	labelSelector := fmt.Sprintf("app=%s", selector)
+	err = WaitForPodReady(ctx, clientset, c.AgnhostNamespace, labelSelector)
+	if err != nil {
+		return fmt.Errorf("error waiting for agnhost pod to be ready: %w", err)
+	}
+
+	return nil
+}
+
+func (c *CreateAgnhostStatefulSet) Prevalidate() error {
+	return nil
+}
+
+func (c *CreateAgnhostStatefulSet) Postvalidate() error {
+	return nil
+}
+
+func (c *CreateAgnhostStatefulSet) getAgnhostDeployment() *appsv1.StatefulSet {
```

Comment:
Also, a ping or curl command with this object should do a exec on this pod to ping or curl given IP or host, that will be nice

Reply:
Currently the objects themselves created by steps aren't saved to the job context. I had a version that exposed struct passing within the job context using map[string]interface{}, but became a nightmare with typecasting and prevalidation. By just keeping parameters in job context, we can do prevalidation ahead of time to make sure everything is populated, and less likely to have "x is nil" passed to a step 1hr into execution 

However, accomplish the mentioned scenario, the parameters passed to the steps themselves are what exist in the job context, see sample scenario in PR #5, the parameter PodName passed to the agnhost create step is then subsequently used in the exec step to curl:

https://github.com/Azure/azure-container-networking/pull/2527/files#diff-4585a7357e679c3568ad5f63a961dc38f34a73087d796e75f718986f60812f5cR37-R60

Open to for future discussion around passing structs to subsequent steps, but then devs need to be very meticulous when authoring, and/or leverage prevalidate steps to write sample objects to context which subsequent steps can also prevalidate